Kubernetes tests are failing in master

Kubernetes tests are failing master build:

Example here:

https://github.com/apache/airflow/actions/runs/350798991


The failing test is:

```
_______________ TestKubernetesPodOperatorSystem.test_pod_failure _______________

self = <kubernetes_tests.test_kubernetes_pod_operator.TestKubernetesPodOperatorSystem testMethod=test_pod_failure>

    def test_pod_failure(self):
        """
        Tests that the task fails when a pod reports a failure
        """
        bad_internal_command = ["foobar 10 "]
        k = KubernetesPodOperator(
            namespace='default',
            image="ubuntu:16.04",
            cmds=["bash", "-cx"],
            arguments=bad_internal_command,
            labels={"foo": "bar"},
            name="test-" + str(random.randint(0, 1000000)),
            task_id="task" + self.get_current_task_name(),
            in_cluster=False,
            do_xcom_push=False,
        )
        with self.assertRaises(AirflowException):
            context = create_context(k)
>           k.execute(context)

kubernetes_tests/test_kubernetes_pod_operator.py:536: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:313: in execute
    status = self.client.read_namespaced_pod(self.name, self.namespace)
.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/api/core_v1_api.py:19078: in read_namespaced_pod
    (data) = self.read_namespaced_pod_with_http_info(name, namespace, **kwargs)  # noqa: E501
.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/api/core_v1_api.py:19169: in read_namespaced_pod_with_http_info
    collection_formats=collection_formats)
.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/api_client.py:345: in call_api
    _preload_content, _request_timeout)
.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/api_client.py:176: in __call_api
    _request_timeout=_request_timeout)
.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/api_client.py:366: in request
    headers=headers)
.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/rest.py:241: in GET
    query_params=query_params)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def request(self, method, url, query_params=None, headers=None,
                body=None, post_params=None, _preload_content=True,
                _request_timeout=None):
        """Perform requests.
    
        :param method: http request method
        :param url: http request url
        :param query_params: query parameters in the url
        :param headers: http request headers
        :param body: request json body, for `application/json`
        :param post_params: request post parameters,
                            `application/x-www-form-urlencoded`
                            and `multipart/form-data`
        :param _preload_content: if False, the urllib3.HTTPResponse object will
                                 be returned without reading/decoding response
                                 data. Default is True.
        :param _request_timeout: timeout setting for this request. If one
                                 number provided, it will be total request
                                 timeout. It can also be a pair (tuple) of
                                 (connection, read) timeouts.
        """
        method = method.upper()
        assert method in ['GET', 'HEAD', 'DELETE', 'POST', 'PUT',
                          'PATCH', 'OPTIONS']
    
        if post_params and body:
            raise ValueError(
                "body parameter cannot be used with post_params parameter."
            )
    
        post_params = post_params or {}
        headers = headers or {}
    
        timeout = None
        if _request_timeout:
            if isinstance(_request_timeout, (int, ) if six.PY3 else (int, long)):  # noqa: E501,F821
                timeout = urllib3.Timeout(total=_request_timeout)
            elif (isinstance(_request_timeout, tuple) and
                  len(_request_timeout) == 2):
                timeout = urllib3.Timeout(
                    connect=_request_timeout[0], read=_request_timeout[1])
    
        if 'Content-Type' not in headers:
            headers['Content-Type'] = 'application/json'
    
        try:
            # For `POST`, `PUT`, `PATCH`, `OPTIONS`, `DELETE`
            if method in ['POST', 'PUT', 'PATCH', 'OPTIONS', 'DELETE']:
                if query_params:
                    url += '?' + urlencode(query_params)
                if re.search('json', headers['Content-Type'], re.IGNORECASE):
                    if headers['Content-Type'] == 'application/json-patch+json':
                        if not isinstance(body, list):
                            headers['Content-Type'] = \
                                'application/strategic-merge-patch+json'
                    request_body = None
                    if body is not None:
                        request_body = json.dumps(body)
                    r = self.pool_manager.request(
                        method, url,
                        body=request_body,
                        preload_content=_preload_content,
                        timeout=timeout,
                        headers=headers)
                elif headers['Content-Type'] == 'application/x-www-form-urlencoded':  # noqa: E501
                    r = self.pool_manager.request(
                        method, url,
                        fields=post_params,
                        encode_multipart=False,
                        preload_content=_preload_content,
                        timeout=timeout,
                        headers=headers)
                elif headers['Content-Type'] == 'multipart/form-data':
                    # must del headers['Content-Type'], or the correct
                    # Content-Type which generated by urllib3 will be
                    # overwritten.
                    del headers['Content-Type']
                    r = self.pool_manager.request(
                        method, url,
                        fields=post_params,
                        encode_multipart=True,
                        preload_content=_preload_content,
                        timeout=timeout,
                        headers=headers)
                # Pass a `string` parameter directly in the body to support
                # other content types than Json when `body` argument is
                # provided in serialized form
                elif isinstance(body, str):
                    request_body = body
                    r = self.pool_manager.request(
                        method, url,
                        body=request_body,
                        preload_content=_preload_content,
                        timeout=timeout,
                        headers=headers)
                else:
                    # Cannot generate the request from given parameters
                    msg = """Cannot prepare a request message for provided
                             arguments. Please check that your arguments match
                             declared content type."""
                    raise ApiException(status=0, reason=msg)
            # For `GET`, `HEAD`
            else:
                r = self.pool_manager.request(method, url,
                                              fields=query_params,
                                              preload_content=_preload_content,
                                              timeout=timeout,
                                              headers=headers)
        except urllib3.exceptions.SSLError as e:
            msg = "{0}\n{1}".format(type(e).__name__, str(e))
            raise ApiException(status=0, reason=msg)
    
        if _preload_content:
            r = RESTResponse(r)
    
            # In the python 3, the response.data is bytes.
            # we need to decode it to string.
            if six.PY3:
                r.data = r.data.decode('utf8')
    
            # log response body
            logger.debug("response body: %s", r.data)
    
        if not 200 <= r.status <= 299:
>           raise ApiException(http_resp=r)
E           kubernetes.client.rest.ApiException: (404)
E           Reason: Not Found
E           HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'Date': 'Sat, 07 Nov 2020 07:33:13 GMT', 'Content-Length': '190'})
E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"test-815505\" not found","reason":"NotFound","details":{"name":"test-815505","kind":"pods"},"code":404}

.build/.kubernetes_venv/lib/python3.6/site-packages/kubernetes/client/rest.py:231: ApiException

```

--- a/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py
+++ b/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py
@@ -310,7 +310,7 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
                 self.log.info("creating pod with labels %s and launcher %s", labels, launcher)
                 final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
             if final_state != State.SUCCESS:
-                status = self.client.read_namespaced_pod(self.name, self.namespace)
+                status = self.client.read_namespaced_pod(self.pod.metadata.name, self.namespace)
                 raise AirflowException(f'Pod returned a failure: {status}')
             return result
         except AirflowException as ex:

--- a/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py
+++ b/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py
@@ -311,7 +311,7 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
                 final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
             if final_state != State.SUCCESS:
                 status = self.client.read_namespaced_pod(self.pod.metadata.name, self.namespace)
-                raise AirflowException(f'Pod returned a failure: {status}')
+                raise AirflowException(f'Pod {self.pod.metadata.name} returned a failure: {status}')
             return result
         except AirflowException as ex:
             raise AirflowException(f'Pod Launching failed: {ex}')

--- a/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
+++ b/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
@@ -190,12 +190,14 @@ class TestKubernetesPodOperator(unittest.TestCase):
         read_namespaced_pod_mock = mock_client.return_value.read_namespaced_pod
         read_namespaced_pod_mock.return_value = failed_pod_status
 
-        with self.assertRaisesRegex(
-            AirflowException,
-            fr'Pod Launching failed: Pod {k.pod.metadata.name} returned a failure: {failed_pod_status}',
-        ):
+        with self.assertRaises(AirflowException) as cm:
             context = self.create_context(k)
             k.execute(context=context)
+
+        self.assertEqual(
+            str(cm.exception), 
+            f"Pod Launching failed: Pod {k.pod.metadata.name} returned a failure: {failed_pod_status}",
+        )
         assert mock_client.return_value.read_namespaced_pod.called
         self.assertEqual(read_namespaced_pod_mock.call_args[0][0], k.pod.metadata.name)
 

--- a/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
+++ b/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
@@ -195,7 +195,7 @@ class TestKubernetesPodOperator(unittest.TestCase):
             k.execute(context=context)
 
         self.assertEqual(
-            str(cm.exception), 
+            str(cm.exception),
             f"Pod Launching failed: Pod {k.pod.metadata.name} returned a failure: {failed_pod_status}",
         )
         assert mock_client.return_value.read_namespaced_pod.called

--- a/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
+++ b/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
@@ -164,3 +164,65 @@ class TestKubernetesPodOperator(unittest.TestCase):
 
         assert start_mock.call_args[0][0].metadata.name.startswith(name_base)
         assert start_mock.call_args[0][0].metadata.name != name_base
+
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.start_pod")
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.monitor_pod")
+    @mock.patch("airflow.kubernetes.kube_client.get_kube_client")
+    def test_describes_pod_on_failure(self, mock_client, monitor_mock, start_mock):
+        from airflow.utils.state import State
+
+        name_base = 'test'
+
+        k = KubernetesPodOperator(
+            namespace='default',
+            image="ubuntu:16.04",
+            cmds=["bash", "-cx"],
+            arguments=["echo 10"],
+            labels={"foo": "bar"},
+            name=name_base,
+            task_id="task",
+            in_cluster=False,
+            do_xcom_push=False,
+            cluster_context='default',
+        )
+        monitor_mock.return_value = (State.FAILED, None)
+        failed_pod_status = 'read_pod_namespaced_result'
+        read_namespaced_pod_mock = mock_client.return_value.read_namespaced_pod
+        read_namespaced_pod_mock.return_value = failed_pod_status
+
+        with self.assertRaises(AirflowException) as cm:
+            context = self.create_context(k)
+            k.execute(context=context)
+
+        self.assertEqual(
+            str(cm.exception), "Pod Launching failed: Pod returned a failure: " + failed_pod_status
+        )
+        assert mock_client.return_value.read_namespaced_pod.called
+        self.assertEqual(read_namespaced_pod_mock.call_args[0][0], k.pod.metadata.name)
+
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.start_pod")
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.monitor_pod")
+    @mock.patch("airflow.kubernetes.kube_client.get_kube_client")
+    def test_no_need_to_describe_pod_on_success(self, mock_client, monitor_mock, start_mock):
+        from airflow.utils.state import State
+
+        name_base = 'test'
+
+        k = KubernetesPodOperator(
+            namespace='default',
+            image="ubuntu:16.04",
+            cmds=["bash", "-cx"],
+            arguments=["echo 10"],
+            labels={"foo": "bar"},
+            name=name_base,
+            task_id="task",
+            in_cluster=False,
+            do_xcom_push=False,
+            cluster_context='default',
+        )
+        monitor_mock.return_value = (State.SUCCESS, None)
+
+        context = self.create_context(k)
+        k.execute(context=context)
+
+        assert not mock_client.return_value.read_namespaced_pod.called

--- a/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
+++ b/tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
@@ -190,13 +190,12 @@ class TestKubernetesPodOperator(unittest.TestCase):
         read_namespaced_pod_mock = mock_client.return_value.read_namespaced_pod
         read_namespaced_pod_mock.return_value = failed_pod_status
 
-        with self.assertRaises(AirflowException) as cm:
+        with self.assertRaisesRegex(
+            AirflowException,
+            fr'Pod Launching failed: Pod {k.pod.metadata.name} returned a failure: {failed_pod_status}',
+        ):
             context = self.create_context(k)
             k.execute(context=context)
-
-        self.assertEqual(
-            str(cm.exception), "Pod Launching failed: Pod returned a failure: " + failed_pod_status
-        )
         assert mock_client.return_value.read_namespaced_pod.called
         self.assertEqual(read_namespaced_pod_mock.call_args[0][0], k.pod.metadata.name)
 

